src/test/java/pages/BasePage.java

```diff
@@ -0,0 +1,138 @@
+    protected final String footerlinksXpath = "//section[@class='footerlinks']";
+    public By searchInputLocator = By.xpath("//input[@id='filter_keyword']");
+    public By searchButtonLocator = By.xpath("//i[@class='fa fa-search']");
+    public By loginAndRegisterButtonLocator = By.xpath("//div[@class='navbar']//a");
+    public By homeLinkByLogoLocator = By.xpath("//a[@class='logo']");
+    public By currencyDropdownLocator = By.xpath("//ul[contains(@class, 'currency')]/parent::*");
+    public By itemsInCartDropDownLocator = By.xpath("//ul[contains(@class, 'topcartopen')]/parent::*");
+    public By newsletterInputLocator = By.xpath(footersocialXpath + "//input[@id='appendedInputButton']");
+    public By newsletterButtonLocator = By.xpath(footersocialXpath + "//button[@class='btn btn-orange']");
+
+    public BasePage(WebDriver driver) {
+        this.driver = driver;
+        int baseWaitInSeconds = Integer.parseInt(ConfigReader.getProperty(ConfigReader.PropertyKeys.BASE_WAIT_IN_SECONDS));
+        this.wait = new WebDriverWait(driver, Duration.ofSeconds(baseWaitInSeconds));
+        this.baseUrl = ConfigReader.getProperty(ConfigReader.PropertyKeys.BASE_URL);
+    }
+
+    public By buttonNavbarLocatorByDataId(NavbarButton button) {
+        return By.xpath(navbarXpath + "//li[@data-id='"+ button.getDataId() +"']//span[@class='menu_text']");
+    }
+
+    public By buttonCategoryMenuLocatorByText(CategoryMenuButton button) {
+        return By.xpath(categoryMenuXpath + "//a[contains(text(), '" + button.getText() + "')]");
+    }
+
+    public By hyperLinkFooterLocatorByText(FooterHyperLink link) {
+        return By.xpath(footerlinksXpath + "//a[text()='" + link.getText() + "']");
+    }
+
+    public void openUrl(String url) {
+        driver.get(url);
+    }
+
+    public void refresh() {
+        driver.navigate().refresh();
+    }
+
+    public void clickOnElement(By locator) {
+        locateElement(locator).click();
+    }
+
+    public void scrollToElement(By locator) {
+        WebElement element = locateElement(locator);
+        JavascriptExecutor js = (JavascriptExecutor) driver;
+        js.executeScript("arguments[0].scrollIntoView(true);", element);
+    }
+
+    public void hoverOverElement(By locator) {
+        WebElement element = locateElement(locator);
+        Actions actions = new Actions(driver);
+        actions.moveToElement(element).perform();
+    }
+
+    public void selectCheckbox(By locator) {
+        WebElement element = locateElement(locator);
+        if (!element.isSelected()) {
+            element.click();
+        }
+    }
+
+    public void deselectCheckbox(By locator) {
+        WebElement element = locateElement(locator);
+        if (element.isSelected()) {
+            element.click();
+        }
+    }
+
+    public void selectRadioButton(By locator) {
+        WebElement element = locateElement(locator);
+        if (!element.isSelected()) {
```

Comment:
I'm always cautious when using `if`s in automated tests - I prefer my tests to be pretty 'silly', there's already too much logic in the backend/frontend implementation :)

Let's imagine:
- you would like to select a radio button
- you assume that the radio button is unselected
- but there's a bug: it is already selected 🐛 
- then our automated test happily skips the conditional and moves on with further testing 🤷‍♀️ 

Maybe we would like the test to fail if the radio button is already selected? Maybe we should throw an exception here? Worth considering :)

Reply:
Interesting observation, I will definitely pay more attention now . I changed the methods that check the state by adding a `throw exception` like this:
```java
    public void selectRadioButton(By locator) {
        WebElement element = locateElement(locator);
        if (element.isSelected()) {
            throw new IllegalStateException("Radio button is already selected");
        }
        element.click();
    }
```